### PR TITLE
[no_early_kickoff] [Train] Recommend `Trainer.restore` on errors raised by `trainer.fit()`

### LINKED ISSUE #33566: [AIR] `TuneError` shouldn't get propagated when using Train only
### What happened + What you expected to happen

When a Train experiment fails during the actual call to `trainer.fit()`, some errors propagated by `tuner.fit()` that gets called under the hood will raise a `TuneError` and tell users to restore using `Tuner.restore`. This should be a training failed error and direct users to use `XTrainer.restore` instead.

```
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "error_prop.py", line 22, in <module>
    t.fit()
  File "/Users/justin/Developer/justinvyu-dev/python/ray/train/base_trainer.py", line 576, in fit
    result_grid = tuner.fit()
  File "/Users/justin/Developer/justinvyu-dev/python/ray/tune/tuner.py", line 352, in fit
    raise TuneError(
ray.tune.error.TuneError: The Ray Tune run failed. Please inspect the previous error messages for a cause. After fixing the issue, you can restart the run from scratch or continue this run. To continue this run, you can use `tuner = Tuner.restore("/Users/justin/ray_results/DataParallelTrainer_2023-03-21_17-22-07")`.
```

### Versions / Dependencies

83dc07a28d7f9427366bc22e7ce57c07af5b63a2

### Reproduction script

```python
import ray
from ray import air, tune
from ray.air import session
from ray.train.data_parallel_trainer import DataParallelTrainer
from ray.tune import Callback


class C(Callback):
    def on_trial_result(self, *args, **kwargs):
        raise RuntimeError()


def train_loop(config):
    session.report({"score": 1})


t = DataParallelTrainer(
    train_loop,
    scaling_config=air.ScalingConfig(num_workers=1),
    run_config=air.RunConfig(callbacks=[C()]),
)
t.fit()

```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

Currently, `TuneError`s that recommend `Tuner.restore` get propagated to users even when they're using `Trainer.fit`. This PR fixes this by wrapping raised errors with a `TrainingFailedError` that includes a message on how to restore the run, and also how to configure a new run to retry on training failures. This PR also separates AIR "error propagation" tests into a new file.

### Notes on the TODOs

I added a few TODOs in the code regarding "driver error propagation" in Tune. Currently, errors that happen in the Tune driver (ex: within user defined callback hooks) get surfaced to the user in **different** ways.

For example:
- An error within `on_trial_result` will surface as a `TuneError(TuneError(OriginalError))`
- An error within `on_checkpoint` will not surface at all. It'll just log a warning.
- An error within `on_step_begin` will surface as `OriginalError`.

This needs to be fixed in 2.5 (all hooks should handle errors in the same way), tracker issue here: TODO.

For the purposes of 2.4, wrapping errors raised by `trainer.fit()` TrainingFailedError error (which already existed) is good enough to fix the issue of "Tune concepts showing up when you're not even using Tune."

## Related issue number

Closes https://github.com/ray-project/ray/issues/33566

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/base_trainer.py [hunk 1]
```diff
@@ -46,7 +46,21 @@
...
+        "your own application logic, but rather an error such as OOM), you can restart "
+        "the run from scratch or continue this run.\n"
+        "To continue this run, you can use: "
+        '`trainer = {trainer_cls_name}.restore("{path}")`.'
+    )
+
+    _FAILURE_CONFIG_MSG = (
+        "To start a new run that will retry on training failures, set "
+        "`air.RunConfig(failure_config=air.FailureConfig(max_failures))` "
+        "in the Trainer's `run_config` with `max_failures > 0`, or `max_failures = 1` "
```
Comment (woshiyyya) [inline python/ray/train/base_trainer.py:61]: Does `max_failures> 0` cover `max_failures=1`?

**Review by woshiyyya — APPROVED** [review]
lgtm, just a few minor comments.

File: python/ray/train/base_trainer.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/base_trainer.py:61]: Oh oops, this should be `max_failures=-1` for infinite retry

**Review by Yard1 — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**